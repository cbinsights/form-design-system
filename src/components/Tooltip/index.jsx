import React from 'react';
import PropTypes from 'prop-types';
import FDS from 'lib/dictionary/js/styleConstants';
import Tippy from '@tippyjs/react';
import { VALID_POSITIONS } from 'components/Popover';

export const DEFAULT_WIDTH = '240px';
export const DELAY_MS = 350;
export const DEFAULT_POSITION = 'bottom';
export const VALID_TEXT_ALIGN = ['left', 'center', 'right'];

const Tooltip = ({
  triggerRef,
  position = DEFAULT_POSITION,
  trigger,
  message,
  maxWidth = DEFAULT_WIDTH,
  interactive,
}) => (
  <Tippy
    arrow={false}
    interactive={interactive}
    delay={[DELAY_MS, 0]}
    reference={triggerRef}
    maxWidth={maxWidth}
    content={message}
    placement={position}
    zIndex={FDS.ZINDEX_MODAL}
    moveTransition="transform 0.2s ease-out"
    appendTo={document.body}
  >
    {trigger}
  </Tippy>
);

Tooltip.propTypes = {
  /** JSX - Acts as a positioning reference for the popover and triggers active state */
  trigger: PropTypes.oneOfType([PropTypes.node, PropTypes.element]).isRequired,

  triggerRef: PropTypes.oneOfType([
    // Either a function
    PropTypes.func,
    // Or the instance of a DOM native element (see the note about SSR)
    PropTypes.shape({ current: PropTypes.any }),
  ]),

  /** Content of the tooltip */
  message: PropTypes.node.isRequired,

  /** Maximum width of tooltip */
  maxWidth: PropTypes.number,

  /**
   * Position preference of tooltip.
   * `top` for example, will place the tooltip above the trigger.
   */
  position: PropTypes.oneOf(VALID_POSITIONS),
  interactive: PropTypes.bool,
};

export default Tooltip;
