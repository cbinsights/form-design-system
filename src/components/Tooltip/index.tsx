import React from 'react';
import FDS from 'dictionary/js/styleConstants';
import Popover, { Position } from 'components/Popover';

export const DEFAULT_WIDTH = '240px';
export const DELAY_MS = 350;
export const DEFAULT_POSITION = 'bottom';

export interface TooltipProps {
  /** JSX - Acts as a positioning reference for the popover and triggers active state */
  trigger: JSX.Element | string;

  /** Content of the tooltip */
  message?: string;

  /** Maximum width of tooltip */
  maxWidth?: number;

  /**
   * Position preference of tooltip.
   * `top` for example, will place the tooltip above the trigger.
   */
  position?: Position;
}

const Tooltip = ({
  trigger,
  message,
  maxWidth,
  position = DEFAULT_POSITION,
}: TooltipProps): JSX.Element => (
  <Popover
    interactionMode="hover"
    trigger={<span style={{ cursor: 'help' }}>{trigger}</span>}
    distance={parseInt(FDS.SPACE_S, 10)}
    position={position}
    alignment="center"
    transitionName="GrowFast"
    delay={DELAY_MS}
  >
    <div
      role="tooltip"
      style={{ maxWidth: maxWidth ? `${maxWidth}px` : DEFAULT_WIDTH }}
      className="bgColor--charcoal inverted align--center padding--y--s padding--x elevation--2 rounded--all fontSize--s fontWeight--bold"
    >
      {message}
    </div>
  </Popover>
);

export default Tooltip;
