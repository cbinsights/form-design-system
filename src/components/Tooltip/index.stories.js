import React from 'react';
import Tooltip from 'components/Tooltip';

const Template = (args) => <Tooltip {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  trigger: <button>ℹ️ Hover Here!</button>,
  message: 'Hello world',
};

export const MaxWidth = Template.bind({});
MaxWidth.args = {
  maxWidth: 600,
  trigger: <button>ℹ️ Hover Here!</button>,
  message:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
};

export default {
  component: Tooltip,
  title: 'components/Tooltip',
};
