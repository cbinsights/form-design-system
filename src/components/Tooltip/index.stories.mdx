import { useRef } from 'react';
import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, optionsKnob as options, text, number } from '@storybook/addon-knobs';
import { arrayToOptions } from 'util/storybook-docs/knobHelpers';

import ImportPath from 'util/storybook-docs/ImportPath';;

import Tooltip, { VALID_TEXT_ALIGN } from 'components/Tooltip';
import { VALID_POSITIONS } from 'components/Popover';

<Meta title="Components/Tooltip" decorators={[withKnobs]} component={Tooltip} />

# Tooltip

<ImportPath component={Tooltip} />

<Story name="Knobs">
  <Tooltip
    trigger={<button>ℹ️ Hover Here!</button>}
    message={text('Message', 'Hello world!')}
    maxWidth={number('Max Width', 160)}
    position={options('Position', arrayToOptions(VALID_POSITIONS), 'bottom', {
      display: 'inline-radio',
    })}
    textAlign={options('Text align', arrayToOptions(VALID_TEXT_ALIGN), 'left', { display: 'inline-radio' })}
  />
</Story>

<Props of={Tooltip} />

## Ref

Supports directly adding a ref (try to only use if you need it)

<Preview>
  <Story name="Ref">
    {() => {
      const ref = useRef();
      return (
        <>
          <button ref={ref}>Trigger</button>
          <Tooltip
            message="Tooltip"
            triggerRef={ref}
          />
        </>
      )
    }}
  </Story>
</Preview>

## Interactive

Specifying interactive allows you to interact with the tooltip even in hover mode

<Preview>
  <Story name="Interactive">
    <Tooltip
      interactive
      trigger={<button>ℹ️ Hover Here!</button>}
      message='Hello world!'
    />
  </Story>
</Preview>

## Max Width

Tooltip comes with a default max width, but this can be customized

<Preview>
  <Story name="Max Width">
    <Tooltip
      interactive
      maxWidth={600}
      trigger={<button>ℹ️ Hover Here!</button>}
      message='Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.'
    />
  </Story>
</Preview>



