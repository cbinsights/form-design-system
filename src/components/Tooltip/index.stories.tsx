import React from 'react';
import { Story } from '@storybook/react';
import Tooltip, { TooltipProps } from 'components/Tooltip';

const Template = (args: TooltipProps) => <Tooltip {...args} />;

export const Primary: Story<TooltipProps> = Template.bind({});
Primary.args = {
  trigger: <button>ℹ️ Hover Here!</button>,
  message: 'Hello world',
};

export const MaxWidth: Story<TooltipProps> = Template.bind({});
MaxWidth.args = {
  maxWidth: 600,
  trigger: <button>ℹ️ Hover Here!</button>,
  message:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
};

export default {
  component: Tooltip,
  title: 'components/Tooltip',
};
