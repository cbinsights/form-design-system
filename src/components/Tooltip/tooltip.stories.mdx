import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, text, number } from '@storybook/addon-knobs';

import { ImportPath } from 'components/util/storybook';

import Tooltip from 'components/Tooltip';

<Meta title="Components/Tooltip" decorators={[withKnobs]} component={Tooltip} />

# Tooltip

<ImportPath component={Tooltip} section="popovers" />

<Story name="Knobs">
  <Tooltip
    trigger={<p>ℹ️Hover Here!</p>}
    message="message"
    maxWidth={number('max width', 240)}
  />
</Story>