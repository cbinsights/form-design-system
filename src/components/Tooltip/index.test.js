import React from 'react';
import { mount } from 'enzyme';

import Tooltip from '.';


const Trigger = () => (<p>trigger</p>);

describe('Tooltip component', () => {
  it('matches snapshot (uses the right classes)', () => {
    const wrapper = mount(<Tooltip trigger={<Trigger />} message="lol" />);
    expect(wrapper).toMatchSnapshot();
  })

});
