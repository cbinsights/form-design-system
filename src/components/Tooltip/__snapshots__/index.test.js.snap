// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component matches snapshot (uses the right classes) 1`] = `
<Tooltip
  message="lol"
  trigger={<Trigger />}
>
  <ForwardRef(TippyWrapper)
    appendTo={<body />}
    arrow={false}
    content="lol"
    delay={
      Array [
        350,
        0,
      ]
    }
    maxWidth={240}
    moveTransition="transform 0.2s ease-out"
    placement="bottom"
    zIndex="20"
  >
    <Tippy
      appendTo={<body />}
      arrow={false}
      content="lol"
      delay={
        Array [
          350,
          0,
        ]
      }
      maxWidth={240}
      moveTransition="transform 0.2s ease-out"
      placement="bottom"
      zIndex="20"
    >
      <Trigger>
        <p>
          trigger
        </p>
      </Trigger>
      <Portal
        containerInfo={
          <div>
            lol
          </div>
        }
      >
        lol
      </Portal>
    </Tippy>
  </ForwardRef(TippyWrapper)>
</Tooltip>
`;
